Metadata API: implement sig verification in Key, store id in key

jku opened the pull request:
Fixes #1417 

Adding id to Key class simplifies life for API users as usually a key needs its identifier: this is visible in how Root.add_key() becomes simpler and in how `verify_signature()` is now possible to implement in key without an additional id argument.

Moving verification to Key arguably also makes the API cleaner because including both "verify myself" and "verify a delegate with threshold" can look awkward in Metadata, and because the somewhat ugly Securesystemslib integration is now Key class implementation detail (see e.g. call to `format_metadata_to_key()`).

Returning bool on verify failure was found to confuse API users (both me and Teodora misused it the first time) and was arguably not a pythonic way to handle it: Now an exception is raised.

There are two issues in this code that are still unresolved:
* #1351: exceptions story is now better but low-level exceptions still bleed through API
* #1422: Metadata should enforce that signatures are unique per key id

I intend to fix these in the next weeks

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Metadata API: Add id to Key

This simplifies life for API users as usually a key needs its
identifier: this is already visible in how update() becomes simpler
in the API.

The downside is that 'from_dict()' now has two arguments (so arguably
the name is not great anymore but it still does _mostly_ the same job
as other from_dicts).

This is an API change, if a minor one.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Move signature verification to Key

This is likely not needed by users of the API (as they are interested
in the higher level functionality "verify delegate metadata with
threshold of signatures").

Moving verify to Key makes the API cleaner because including both
"verify myself" and "verify a delegate with threshold" can look awkward
in Metadata, and because the ugly Securesystemslib integration is now
Key class implementation detail (see Key.to_securesystemslib_key()).

Also raise on verify failure instead of returning false: this was found
to confuse API users (and was arguably not a pythonic way to handle it).

* Name the function verify_signature() to make it clear what is being
  verified.
* Assume only one signature per keyid exists: see #1422
* Raise only UnsignedMetadataError (when no signatures or verify failure),
  the remaining lower level errors will be handled in #1351
* Stop using a "keystore" in tests for the public keys: everything we
  need is in metadata already

This changes API, but also should not be something API users want to
call in the future when "verify a delegate with threshold" exists.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

sechkova commented on tuf/api/metadata.py:
```diff
@@ -417,6 +417,7 @@ class Key:
     """A container class representing the public portion of a Key.
 
     Attributes:
+        id: An identifier string
```
"An identifier string" is true from the Key class perspective but I think some hints about it being the digest of  an ambiguous canonical key is needed.

jku replied:
yeah, makes sense -- I'm not fond of documenting the (probably untrue) idea that we could calculate the id from the key content but I should document that the id must be unique within the keys in this Metadata

jku replied:
```
        id: An identifier string that must uniquely identify a key within
            the metadata it is used in. This implementation does not verify
            that the id is the hash of a specific representation of the key.
```
this work for you?

sechkova replied:
works for me :)

sechkova commented on tests/test_api.py:
```diff
@@ -162,50 +159,46 @@ def test_read_write_read_compare(self):
+
+        # Locate the public keys we need from root
+        targets_keyid = next(iter(root.roles["targets"].keyids))
+        targets_key = root.keys[targets_keyid]
```
I noticed now this line in the spec saying

> ... Clients MUST calculate each KEYID to verify this is correct for the associated key ...

Calculating the `KEYID` is now done during `sslib.format_metadata_to_key` but is never compared to the one read from the metadata files. I can't decide is there a benefit in explicitly doing this check. Or is it done somewhere and I am missing it?

jku replied:
* it's not done anywhere
* I've read through the sslib implementation and I don't think the keyid calculation is actually reproducible -- things like securesystemslib.settings.HASH_ALGORITHMS value affect it.
* I think specifying how the unique key must be created in the spec is a mistake: it serves no purpose. The spec should just demand the keyid uniquely identifies the key within this metadata file
* I also think that's how we should handle keyid in python-tuf, even if spec is what it is: IMO we can't do anything else

jku replied:
see also https://github.com/theupdateframework/taps/blob/master/tap12.md

sechkova replied:
Thanks for the explanation. Feel like I am convinced now!

joshuagl replied:
> I noticed now this line in the spec saying
> 
> > ... Clients MUST calculate each KEYID to verify this is correct for the associated key ...
> 
> Calculating the `KEYID` is now done during `sslib.format_metadata_to_key` but is never compared to the one read from the metadata files. I can't decide is there a benefit in explicitly doing this check. Or is it done somewhere and I am missing it?

FYI https://github.com/theupdateframework/tuf/pull/1014, specifically https://github.com/theupdateframework/tuf/commit/9fa92e22865b2ba4e164493564042cf3ca778bee

sechkova reviewed (commented):
As already mentioned in the issue, I am in favour of these changes. 
Left some thoughts/questions as comments.

sechkova commented on tuf/api/metadata.py:
```diff
@@ -458,6 +408,51 @@ def to_dict(self) -> Dict[str, Any]:
+
+            signed_serializer = CanonicalJSONSerializer()
+
+        sslib_key, dummy = sslib_keys.format_metadata_to_key(self.to_dict())
```
Given the latest discovery of the correct usage of `sslib_keys.format_metadata_to_key` in this use case, it seems that it needs to receive `default_keyid` so that it does not recalculate the `keyid`.

sechkova replied:
or not using it at all ... :roll_eyes:

jku replied:
yeah good find: I think I'll try dropping the whole call, that function does not look like something I'd like to call

jku commented:
Changes based on Teodoras review:
* Added a bit of docstring on key id
* Removed the call to `securesystemslib.keys.format_metadata_to_key()`, added `Key.to_securesystemslib_key()` and used that instead

joshuagl reviewed (commented):
I know that it feels a bit repetitive/redundant in the API, but given the consistent use of the term `keyid` in the specification and in the code (i.e. `Key.__init__()` and `Key.from_dict()` both take a `keyid` parameter), would it make sense to rename `Key.id` -> `Key.keyid`?

jku commented:
> I know that it feels a bit repetitive/redundant in the API, but given the consistent use of the term `keyid` in the specification and in the code (i.e. `Key.__init__()` and `Key.from_dict()` both take a `keyid` parameter), would it make sense to rename `Key.id` -> `Key.keyid`?

I would have called the arguments id as well but that shadows the builtin `id()` which feels wrong... so I guess with the spec-consistency argument that means you have a point: consistent use of keyid would make most sense

joshuagl reviewed (approved):
Other than id -> keyid, this LGTM.

jku commented:
oh forgot to metnion: the only changes in the latest force push are renaming "key.id" to "key.keyid"
